Rob/1.36.0 docs (#1314)
FusionAuth/fusionauth-site: site/docs/v1/tech/apis/scim/_scim-enterpriseuser-request-body.adoc

include::/docs/v1/tech/apis/scim/_scim-common-user-header.adoc[]

The following is an example of a typical SCIM User create request body a SCIM client might send. However, your incoming request lambda could use these values in completely different ways to populate the FusionAuth User object if you choose.

This is taken from the SCIM Schema RFC describing a SCIM EnterpriseUser schema. For the full specification you can find the RFC link:https://datatracker.ietf.org/doc/html/rfc7643#section-4.3[here]

==== Request Body
[.api]
include::/docs/v1/tech/apis/scim/_scim-common-body.adoc[]

include::/docs/v1/tech/apis/scim/_scim-common-user-request-body.adoc[]

[field]#urn:ietf:params:scim:schemas:extension:enterprise:2.0:User# [type]#[String]# [optional]#Required#::
This is the root property of the EnterpriseUser schema extension. It should match the identifier included in the schemas property.

[field]#urn:ietf:params:scim:schemas:extension:enterprise:2.0:User.employeeNumber# [type]#[String]# [optional]#Optional#::

> A string identifier, typically numeric or alphanumeric, assigned to a person, typically based on order of hire or association with an organization.

Author: Yes, it's just an extension of the SCIM User schema per their spec.

[field]#urn:ietf:params:scim:schemas:extension:enterprise:2.0:User.costCenter# [type]#[String]# [optional]#Optional#::
Identifies the name of a cost center.

[field]#urn:ietf:params:scim:schemas:extension:enterprise:2.0:User.organization# [type]#[String]# [optional]#Optional#::
Identifies the name of an organization.

[field]#urn:ietf:params:scim:schemas:extension:enterprise:2.0:User.division# [type]#[String]# [optional]#Optional#::
Identifies the name of a division.

[field]#urn:ietf:params:scim:schemas:extension:enterprise:2.0:User.department# [type]#[String]# [optional]#Optional#::
Identifies the name of a department.

[field]#urn:ietf:params:scim:schemas:extension:enterprise:2.0:User.manager# [type]#[String]# [optional]#Optional#::

> The user's manager.  A complex type that optionally allows service providers to represent organizational hierarchy by referencing the "id" attribute of another User.

Author: FusionAuth is the service provider, but we don't have the concept of hierarchies like that so it would be up to the dev to figure out what they want to do with that data.

[field]#urn:ietf:params:scim:schemas:extension:enterprise:2.0:User.manager.value# [type]#[String]# [optional]#Optional#::

> The "id" of the SCIM resource representing the user's manager

Author: We don't really have the concept of manager so it's up to the dev what they want it to be really, but most likely it would be the FA User Id of the "manager"

> Because SCIM schemas are completely customizable, There are no required fields for the incoming request other than the common fields that all SCIM requests must include to match the specification. The only requirement about the user data is that in the incoming request lambda, you must specify either the **email** or the **username** or both for the FusionAuth User object. Either of these values may be used to uniquely identify the User and may be used to authenticate the User. Otherwise, it is completely up to your defined lambda to match any other incoming user data to the appropriate FusionAuth User fields for your configuration.

Reviewer: ```suggestion
Because SCIM schemas are completely customizable, there are no required fields for the incoming request other than the common fields that all SCIM requests must include to match the specification.
The only requirement about the user data is that in the incoming request lambda, you must specify either the **email**, the **username**, or both for the FusionAuth User object.
Either of these values may uniquely identify the User and can be used during a login.
For all other fiends, it is completely up to your lambda to map incoming user data to the appropriate FusionAuth User fields.
```

Reviewer: If you have multiple lines of text with no newlines between them, Asciidoc collapses them to a paragraph. Multiple lines are nice for diffs.

> The following is an example of a typical SCIM User create request body your SCIM client might send. However, your incoming request lambda could use these values in completely different ways to populate the FusionAuth User object if you choose.

Reviewer: I'm confused about the "however" sentence. Seems like it belongs later in the process because here we are talking about the input before the mapping, right?

> This is taken from the SCIM Schema RFC describing a SCIM User schema. For the full specification you can find the RFC here:

Reviewer: ```suggestion
This is taken from the SCIM Schema RFC describing a SCIM User schema. https://datatracker.ietf.org/doc/html/rfc7643#page-19[Read the full specification here].
```

Reviewer: Also, point to a section, not a page

> - https://datatracker.ietf.org/doc/html/rfc7643#page-19

Reviewer: You could just put a link rather than have a sep bullet point.

> ==== Request Body

Reviewer: You kind of make this serve two purposes:

* a sample request body, from the spec
* suggested mappings to the FusionAuth user object

I'd suggest doing just the first, and then building a table that offers a typical mapping. This could be a different section of the API docs.

[.api]
[field]#active# [type]#[Boolean]# [required]#Required#::
The state of the SCIM Resource.

[field]#schemas# [type]#[Array]# [required]#Required#::

> The String identifier of the SCIM Resource schemas and extensions provided in the request. This should match supported schemas and extensions for the provisioning server.

Reviewer: Do we only handle certain schemas?

Author: Yes and no. Technically they can handle whatever they want in the Lambda as long as it maps the incoming schema values to FusionAuth concepts like User and Group. They are free the advertise that they support whatever schemas and customizations they want, but ultimately they are responsible for converting the requests and responses in their lambdas.

[field]#externalId# [type]#[String]# [optional]#Optional#::
The externalId is a unique identifier created by the client that identifies the resource in the SCIM client's environment. It could be used as the FusionAuth resource id as well, but that is not a requirement of the SCIM specification. That would need to be done by the defined incoming lambda if desired.

[field]#name.formatted# [type]#[String]# [optional]#Optional#::

> his would typically be the value used to populate the FusionAuth User.fullName by the incoming request lambda.

Reviewer: ```suggestion
This would typically be the value used to populate the FusionAuth User.fullName by the incoming request lambda.
```

Reviewer: The `T` is missing for many of these fields.

[field]#name.familyName# [type]#[String]# [optional]#Optional#::
his would typically be the value used to populate the FusionAuth User.lastName by the incoming request lambda.

[field]#name.givenName# [type]#[String]# [optional]#Optional#::
his would typically be the value used to populate the FusionAuth User.firstName by the incoming request lambda.

[field]#name.middleName# [type]#[String]# [optional]#Optional#::
his would typically be the value used to populate the FusionAuth User.middleName by the incoming request lambda.

[field]#name.honorificPrefix# [type]#[String]# [optional]#Optional#::
A prefix to be applied to the user's full name.

[field]#name.honorificSuffix# [type]#[UUID]# [optional]#Optional#::
A suffix to be applied to the user's full name.

[field]#userName# [type]#[String]# [optional]#Optional#::
This would typically be the value used to populate the FusionAuth User.userName by the incoming request lambda.

[field]#password# [type]#[String]# [optional]#Optional#::
This would typically be the value used to populate The FusionAuth User.password by the incoming request lambda.

[field]#phoneNumbers# [type]#[Array]# [optional]#Optional#::

> SCIM supports multiple phone numbers while FusionAuth only allows a single phone number for a user. The incoming request lambda would need to determine which one to use if multiple are specified. The most common strategy would be to just use the one where primary is true.

Reviewer: `FusionAuth only allows a single phone number for a user` -> `FusionAuth only has a single field, mobilePhone. Additional phone numbers can be stored in the data field`

[field]#phoneNumbers[x].primary# [type]#[Boolean]# [optional]#Optional#::
Indicates that this phone number is the user's primary phone number.

[field]#phoneNumbers[x].value# [type]#[String]# [optional]#Optional#::
This would typically be used to populate the FusionAuth User.mobilePhone by the incoming request lambda.

[field]#phoneNumbers[x].type# [type]#[String]# [optional]#Optional#::

> This might be used to alternatively choose a value to decide which phone number to use to populate the FusionAuth User.mobilePhone by the incoming request lambda.

Reviewer: ```suggestion
The type of the phone number (work, home, mobile). Rather than use [field]#phoneNumbers[x].primary#, you could select a different number as the FusionAuth User.mobilePhone.
```
